### PR TITLE
Support native S3 conditional writes

### LINKED ISSUE #6285: object_store: Conditional put and rename_if_not_exist on S3
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

This is now supported on S3: https://aws.amazon.com/about-aws/whats-new/2024/08/amazon-s3-conditional-writes/

**Describe the solution you'd like**

We should make sure this works out-of-the-box. Also need to maintain compatibility with Minio, Localstack, and CloudFlare R2.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by benesch**
Add support for `PutMode::Create` and `copy_if_not_exists` on native AWS S3, which uses the underlying conditional write primitive that Amazon launched earlier this year [0].

The conditional write primitive is simpler than what's available in other S3-like products (e.g., R2), so new modes for `s3_copy_if_not_exists` and `s3_conditional_put` are added to select the native S3-specific behavior.

To maintain strict backwards compatibility (e.g. with older versions of LocalStack), the new behavior is not on by default. It must be explicitly requested by the end user.

The implementation for `PutMode::Create` is straightforward. The implementation of `copy_if_not_exists` is a bit more involved, as it requires managing a multipart upload that uses the UploadPartCopy operation, which was not previously supported by this crate's S3 client.

To ensure test coverage, the object store workflow now runs the AWS integration tests with conditional put both disabled and enabled.

[0]: https://aws.amazon.com/about-aws/whats-new/2024/08/amazon-s3-conditional-writes/

# Which issue does this PR close?

Fix #6285.

### CODE REVIEW

**Comment by benesch** [comment]
@tustvold are you the right person to review this? I saw you implemented quite a bit of the previous conditional put support (via DynamoDB).

**Comment by benesch** [comment]
> To maintain strict backwards compatibility (e.g. with older versions of LocalStack), the new behavior is not on by default. It must be explicitly requested by the end user.

This isn't ideal, but it seemed best for the short term to avoid breaking backcompat for anyone who might be using a version of S3 that doesn't support CAS. In the long term, I think the ideal would definitely be to have the native S3 CAS support enabled by default (i.e. unless overridden explicitly by the user).

File: object_store/src/aws/mod.rs [hunk 1]
```diff
@@ -293,6 +297,34 @@ impl ObjectStore for AmazonS3 {
...
+                    .client
+                    .complete_multipart(
+                        to,
+                        &upload_id,
+                        vec![part_id],
+                        CompleteMultipartMode::Create,
+                    )
+                    .await
+                {
+                    Err(e @ Error::Precondition { .. }) => Err(Error::AlreadyExists {
```
Comment (tustvold) [inline object_store/src/aws/mod.rs:319]: Do we need to abort the multipart upload in this case? Does this happen automatically?

File: object_store/src/aws/precondition.rs [hunk 2]
```diff
@@ -46,6 +46,15 @@ pub enum S3CopyIfNotExists {
     ///
     /// Encoded as `header-with-status:<HEADER_NAME>:<HEADER_VALUE>:<STATUS>` ignoring whitespace
     HeaderWithStatus(String, String, reqwest::StatusCode),
+    /// Native Amazon S3 supports copy if not exists through a multipart upload
+    /// where the upload copies an existing object and is completed only if
+    /// the new object does not already exist.
+    ///
+    /// WARNING: When using this mode, `copy_if_not_exists` does not copy
+    /// tags or attributes from the source object.
+    ///
```
Comment (tustvold) [inline object_store/src/aws/precondition.rs:55]: Could possibly also link to the module docs around lifecycle rules for cleaning up incomplete multipart uploads - https://docs.rs/object_store/latest/object_store/aws/index.html#multipart-uploads

**Review by tustvold — APPROVED** [review]
Thank you, this looks good to me, mostly just some minor nits.

I agree with the decision to leave this off by default for a couple of reasons:

* S3 compatible stores support these mechanisms with fewer drawbacks (e.g. full conditional update, native copy_if_not_exists, etc...) and users should prefer these
* AWS may _eventually_ bring S3 into feature parity with literally every other object store, at which point this will provide a coherent API to converge on

File: object_store/src/aws/client.rs [hunk 3]
```diff
@@ -99,7 +99,10 @@ pub(crate) enum Error {
     CreateMultipartResponseBody { source: reqwest::Error },
 
     #[snafu(display("Error performing complete multipart request: {}", source))]
-    CompleteMultipartRequest { source: crate::client::retry::Error },
+    CompleteMultipartRequest {
+        source: crate::client::retry::Error,
+        path: String,
```
Comment (calavera) [inline object_store/src/aws/client.rs:104]: should the path also be in the displayed error?

File: object_store/src/aws/mod.rs (hunk 1 above)
Comment (benesch) [reply object_store/src/aws/mod.rs]: Oof, good call. It does not happen automatically. I added code to do a best effort multipart upload, and a docs note that this isn't guaranteed and that the user should configure automatic multipart expiration via a lifecycle rule.

**Review by benesch — COMMENTED** [review]
Thanks very much for the quick review, @tustvold! Addressed all your feedback, I believe.

**Comment by benesch** [comment]
@tustvold if you can give me another re-run here I think we’ll be all green on CI. I fixed a minor clippy failure and upgraded to a version of localstack that supports conditional put.

**Comment by benesch** [comment]
Whew, all tests here are green! @tustvold let me know if there's anything else here you'd like to see, but from my perspective this is ready to (squash) merge.

File: object_store/src/aws/mod.rs [hunk 4]
```diff
@@ -293,6 +297,47 @@ impl ObjectStore for AmazonS3 {
...
+                let upload_id = self
+                    .client
+                    .create_multipart(to, PutMultipartOpts::default())
+                    .await?;
+
+                let res = async {
+                    let part_id = self
+                        .client
+                        .put_part(to, &upload_id, 0, PutPartPayload::Copy(from))
+                        .await?;
```
Comment (tustvold) [inline object_store/src/aws/mod.rs:310]: I think this error should also trigger cleanup, might be worth encapsulating the steps into a function

**Review by tustvold — COMMENTED** [review]
Just a small nit, then I think we're good to go. Thank you for this

File: object_store/src/aws/mod.rs (hunk 4 above)
Comment (benesch) [reply object_store/src/aws/mod.rs:310]: It does, doesn't it? If anything in this `async` block (L306-329) returns an error, the cleanup on L336 should fire.

Comment (tustvold) [reply object_store/src/aws/mod.rs:310]: Oh I missed that this is wrapped in an async block, cunning

**Comment by criccomini** [comment]
Amazing. Any idea when this might go out 🔥

**Comment by tustvold** [comment]
https://github.com/apache/arrow-rs/issues/6596 tracks the next release

File: object_store/src/aws/mod.rs (hunk 4 above)
Comment (benesch) [reply object_store/src/aws/mod.rs:310]: The fact that async blocks can function like [try blocks](https://doc.rust-lang.org/beta/unstable-book/language-features/try-blocks.html) is one of my favorite Rust tricks!

**Comment by benesch** [comment]
Thanks very much for the review and merge, @tustvold!

File: object_store/src/aws/precondition.rs [hunk 5]
```diff
@@ -118,6 +138,17 @@ pub enum S3ConditionalPut {
...
+    /// Like `ETagMatch`, but with support for `PutMode::Create` and not
+    /// `PutMode::Option`.
+    ///
+    /// This is the limited form of conditional put supported by Amazon S3
+    /// as of August 2024 ([announcement]).
+    ///
+    /// Encoded as `etag-put-if-not-exists` ignoring whitespace.
+    ///
+    /// [announcement]: https://aws.amazon.com/about-aws/whats-new/2024/08/amazon-s3-conditional-writes/
+    ETagPutIfNotExists,
```
Comment (tustvold) [inline object_store/src/aws/precondition.rs:150]: S3 now supports the full header so I wonder if we need this anymore?

https://github.com/apache/arrow-rs/issues/6799

**Pull request merged**